### ISSUE TITLE
(.. | numbers) += 1 fails with: "left hand side must be FieldAccess"

### ISSUE THREAD

**Issue by btiernay** [eiiches/jackson-jq#7]
The following code snippet:

``` java
    val input = DEFAULT.createObjectNode();
    input.put("x", 1);

    val expression = "(.. | numbers) += 1";
    val query = JsonQuery.compile(expression);

    val result = query.apply(input).get(0);
```

fails with:

```
net.thisptr.jackson.jq.exception.IllegalJsonArgumentException: left hand side must be FieldAccess
    at net.thisptr.jackson.jq.internal.tree.binaryop.assignment.ComplexAssignment.apply(ComplexAssignment.java:31)
    at net.thisptr.jackson.jq.internal.IsolatedScopeQuery.apply(IsolatedScopeQuery.java:21)
    at net.thisptr.jackson.jq.JsonQuery.apply(JsonQuery.java:20)

```

Please see http://stackoverflow.com/questions/35609548/how-to-apply-a-function-to-all-strings-in-records-structure-recursively-using-j for context.

**Comment by btiernay** [comment]
Any ideas why this might be happening? I'd be willing to help create a fix if I could get some pointers. Thanks!

**Comment by eiiches** [comment]
@btiernay I'm sorry to have kept you waiting, I'll try to look closely into the problem this weekend.

**Comment by eiiches** [comment]
Sorry for my late response, I'm currently working on a fix.

The problem was that we cannot determine which part of an input JsonNode to update because in the current implementation we only handled a simple FieldAccess like `.a[0].a` in assignments.

In btiernay's example, we have an input `{"x": 1}` and `(..|numbers)` evaluates to `1`. Then we apply `+= 1` against `1` and get `2`. But where in `{"x": 1}` we should put this `2` back, we do not know now.

I'm trying to fix this by creating a simple wrapper around JsonNode and wrapping an input before evaluating assignment expressions. So far `(..|numbers) += 1` worked, but test cases (such as `jq '.[1] = 1' <<< '[]'`, in which a non-existent element of an input array or object is accessed) started to fail and still working on a fix.

**Comment by btiernay** [comment]
Thanks for your support, much appreciated!

**Comment by mdbhat** [comment]
I am facing a similar issue. Do we have a solution for this? My rule works fine in jqplay but I get this error in java.
net.thisptr.jackson.jq.exception.IllegalJsonArgumentException: left hand side must be FieldAccess
	at net.thisptr.jackson.jq.internal.tree.binaryop.assignment.UpdateAssignment.apply(UpdateAssignment.java:28)
	at net.thisptr.jackson.jq.internal.tree.PipedQuery.applyRecursive(PipedQuery.java:42)
	at net.thisptr.jackson.jq.internal.tree.PipedQuery.apply(PipedQuery.java:25)
	at net.thisptr.jackson.jq.internal.IsolatedScopeQuery.apply(IsolatedScopeQuery.java:21)
	at net.thisptr.jackson.jq.JsonQuery.apply(JsonQuery.java:20)

**Comment by mdbhat** [comment]
It worked for me by making this change in the rule
(. |= (. + if has("dataSource") then {("newDataSource"): .dataSource} else . end)) | del(.dataSource)
New rule:
(. + if has("dataSource") then {("newDataSource"): .dataSource} else . end) | del(.dataSource)

**Comment by eiiches** [comment]
This long-standing issue has been resolved in 1.0.0-preview.* versions.
